### PR TITLE
1.9.0 dev

### PR DESCRIPTION
**Pull request by nself**
Implements worker-master monitoring for simp-tsds, other refactoring for bugfixes.

### CODE REVIEW

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 1]
```diff
@@ -414,58 +461,122 @@ sub _create_worker
...
+            $self->logger->error(sprintf("Could not find a status file for worker %s", $worker->worker_name));
+            $errors_found += 1;
+        }
+    }
+    $self->logger->debug(sprintf("simp-tsds found %s of %s workers reporting errors or with stale/missing status files in", $errors_found, scalar(keys(%{$self->worker_patterns}))));
+    my $res = write_service_status(
+        service_name    => 'simp-tsds',
+        error           => ($errors_found > 0) + 0,
+        error_txt       => ($errors_found > 0 ? 
+            sprintf("simp-tsds found %s of %s workers reporting errors or with stale/missing status files in",
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Master.pm:571]: Move this `sprintf()` out of and above the `$res` assignment and assign its value to a new string variable.
Use that string variable in the debug logging currently on line 566.
Use it again in the error_txt parameter here.
Use it again in a new error logging line if `$errors_found > 0`.

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 2]
```diff
@@ -414,58 +461,122 @@ sub _create_worker
...
+                $errors_found,
+                scalar(keys(%{$self->worker_patterns})),
+                $self->worker_status_dir) : ""),
+    );
+    if (!$res) {
+        # Should this cause an exit(), since monitoring will be broken?
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Master.pm:577]: We will rely upon status file freshness checks to catch this issue.

File: lib/GRNOC/Simp/TSDS/Pusher.pm [hunk 3]
```diff
@@ -53,6 +53,7 @@ has tsds_config => (
 
 has tsds_svc => (is => 'rwp');
 
+
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Pusher.pm:56]: Nitpick: Remove extra newline unless it was intentional for readability

File: lib/GRNOC/Simp/TSDS/Worker.pm [hunk 4]
```diff
@@ -404,9 +423,36 @@ sub _push_data {
     $self->logger->debug($self->worker_name . " Sending " . scalar(@{$self->messages}) . " messages");
 
     my $iterator = natatime(100, @{$self->messages});
+    my $last_failure;
+    my $res;
+    while (my @block = $iterator->()) {
+	    $res = $self->tsds_pusher->push(\@block);
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Worker.pm:429]: Remove extra tab. Assignment is indented more than the rest of the code block.

**Review by veorlo — COMMENTED** [review]
There are a number of small changes that would be nice to have before we merge and release. Everything looks good as far as functionality.

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 5]
```diff
@@ -93,6 +115,8 @@ has run_group => (
 
 =item children
 
+=item worker_patterns
+
 =item hup
```
[the remark this answers is not in the document]
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:120]: Removing.

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 6]
```diff
@@ -205,11 +231,20 @@ sub start
     # Only run once unless HUP gets set, then reload and go again
```
[the remark this answers is not in the document]
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:231]: Commented

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 7]
```diff
@@ -355,10 +403,9 @@ sub _create_workers
             $self->logger->info("Child $pid has exited.");
         }
 
-        $running->send;
+        $self->running->send;
```
[the remark this answers is not in the document]
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:406]: Commented.

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 8]
```diff
@@ -384,13 +431,13 @@ sub _create_collection_workers
...
     {
-        my $worker_name = "$collection->{'composite'} [$worker_id]";
+        my $worker_name = sprintf("%s [%s]", $collection->{'composite'}, $worker_id);
 
         $self->_create_worker(
             name       => $worker_name,
             collection => $collection,
             hosts      => $worker_hosts{$worker_id},
-	    offset     => $offset++
+	        offset     => $offset++
```
[the remark this answers is not in the document]
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:440]: Leaving as is; this renders as properly formatted on my host, and in github. Previously, this line was lacking an indent. I think the way it's displayed in GitHub's diff has made it appear off?

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 9]
```diff
@@ -414,58 +461,122 @@ sub _create_worker
     # slower to start up
     my $actual_offset = ($stagger_offset * $stagger) % $interval;
 
+    my $required_vals =
+        defined $collection->{required_values}
+        ? $collection->{'required_values'}
+        : '';
+
+    my $excludes =
+        defined $collection->{'exclude'} ? $collection->{exclude} : [];
```
[the remark this answers is not in the document]
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:470]: Made this and the above statement one-liners.

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 10]
```diff
@@ -414,58 +461,122 @@ sub _create_worker
...
+                # Check that status file has a clear error flag i.e. 0
+                if ( $worker_status->{'error'} ne 0 ) {
+                    $self->logger->error(sprintf("Non-zero error flag from %s. Error: %s",
+                        $worker->worker_name,
+                        $worker_status->{'error_text'})
+                    );
+                    $errors_found += 1; 
+                } 
+                # Check that timestamp is fresh (timestamp within 2 * interval of current time)
+                if ( $worker_status->{'timestamp'} < (time() - $worker->interval * 2) ) {
```
[the remark this answers is not in the document]
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:547]: Implemented.

File: lib/GRNOC/Simp/TSDS/Master.pm (hunk 1 above)
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:571]: Implemented.

File: lib/GRNOC/Simp/TSDS/Master.pm (hunk 2 above)
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:577]: Comment removed.

File: lib/GRNOC/Simp/TSDS/Master.pm (hunk 8 above)
[the remark this answers is not in the document]
Comment (veorlo) [reply lib/GRNOC/Simp/TSDS/Master.pm:440]: Ah, there's a `\t` char at the start of that line. You can see it here if it's highlighted. This was definitely due to improper tabulation standards at the time this was created.

File: lib/GRNOC/Simp/TSDS/Pusher.pm (hunk 3 above)
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Pusher.pm:56]: Removed.

File: lib/GRNOC/Simp/TSDS/Worker.pm (hunk 4 above)
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Worker.pm:429]: This is a tab-spaces mixing issue. All indentation for these three files has been converted to spaces in the final commit for this PR

File: lib/GRNOC/Simp/TSDS/Master.pm [hunk 11]
```diff
@@ -563,18 +560,14 @@ sub _check_worker_health() {
             $errors_found += 1;
         }
     }
-    $self->logger->debug(sprintf("simp-tsds found %s of %s workers reporting errors or with stale/missing status files in", $errors_found, scalar(keys(%{$self->worker_patterns}))));
+    my $status_message = sprintf("simp-tsds found %s of %s workers reporting errors or with stale/missing status files in", $errors_found, scalar(keys(%{$self->worker_patterns})));
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Master.pm:563]: Add ` %s` at the end of the format string and add the `$self->worker_status_dir` to the args of `sprintf()` to help identify the directory being used.

File: lib/GRNOC/Simp/TSDS/Worker.pm [hunk 12]
```diff
@@ -276,74 +276,74 @@ sub _setup_worker {
     # Create polling timer for event loop
     while (1){
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Worker.pm:277]: All of the code inside this while loop needs to be indented one more time after auto-removal got rid of `\t` chars used for indentation. At least, it looks like the while loop's code block is at the same indentation level as its declaration.

File: lib/GRNOC/Simp/TSDS/Worker.pm [hunk 13]
```diff
@@ -276,74 +276,74 @@ sub _setup_worker {
...
+        $self->logger->error($self->worker_name." Disconnected from RabbitMQ, restarting");
+        return;
+    }
+    
+    # Pull data for each host from Comp
+    for my $host (@{$self->hosts}) {
+        
+        $self->logger->info($self->worker_name." Processing $host");
+
+        my %args = (
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Worker.pm:296]: Add indentation for the KVPs now that `\t` was removed.

File: lib/GRNOC/Simp/TSDS/Worker.pm [hunk 14]
```diff
@@ -276,74 +276,74 @@ sub _setup_worker {
...
+        time           => $now - $self->interval 
+        );
+        
+        # if we're trying to only get a subset of values out of
+        # simp, add those arguments now. This presumes that SIMP
+        # and SIMP-collector have been correctly configured to have
+        # the data available validity is checked for earlier
+        # in Master
+        if ($self->filter_name) {
+        $args{$self->filter_name} = $self->filter_value;
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Worker.pm:311]: Add indent

File: lib/GRNOC/Simp/TSDS/Worker.pm [hunk 15]
```diff
@@ -276,74 +276,74 @@ sub _setup_worker {
...
+        }
+        
+        # to provide some degree of backward compatibility,
+        # we only put this field on if we need to:
+        if (scalar(@{$self->exclude_patterns}) > 0) {
+        $args{'exclude_regexp'} = $self->exclude_patterns;
```
Comment (veorlo) [inline lib/GRNOC/Simp/TSDS/Worker.pm:317]: Add indent

**Review by veorlo — COMMENTED** [review]
A few small changes need to be made and then this can be merged!

File: lib/GRNOC/Simp/TSDS/Master.pm (hunk 11 above)
Comment (nself) [reply lib/GRNOC/Simp/TSDS/Master.pm:563]: Good catch, fixed.

**Review by veorlo — APPROVED** [review]
Looks great!

**Pull request merged**